{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finalize the single cell perturbseq dataset\n",
    "\n",
    "The output of the Seurat pipeline in 2.process-perturbseq.ipynb is not easily compatible with our downstream tasks.\n",
    "This notebook finalizes the input perturbseq (CRISPRi) dataset.\n",
    "\n",
    "There are four basic steps:\n",
    "\n",
    "1. Load, transpose, and clean the gene expression matrix\n",
    "2. Load the single cell file identities (barcodes and metadata)\n",
    "3. Merge\n",
    "4. Output\n",
    "\n",
    "Lastly, we use pycytominer.aggregate to form bulk profiles from the single cell readouts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from pycytominer import aggregate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gse_id = \"GSE132080\"\n",
    "perturbseq_data_dir = pathlib.Path(\"data/perturbseq/\")\n",
    "\n",
    "output_file = pathlib.Path(f\"{perturbseq_data_dir}/{gse_id}_final_analytical.tsv.gz\")\n",
    "output_bulk_file = pathlib.Path(f\"{perturbseq_data_dir}/{gse_id}_bulk_final_analytical.tsv.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(23537, 1002)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metadata_barcode</th>\n",
       "      <th>Metadata_sequence</th>\n",
       "      <th>ABCA1</th>\n",
       "      <th>ABCC3</th>\n",
       "      <th>ABI3BP</th>\n",
       "      <th>AC002331.1</th>\n",
       "      <th>AC002480.3</th>\n",
       "      <th>AC003092.1</th>\n",
       "      <th>AC005616.2</th>\n",
       "      <th>AC006262.5</th>\n",
       "      <th>...</th>\n",
       "      <th>YPEL4</th>\n",
       "      <th>YPEL5</th>\n",
       "      <th>ZBTB38</th>\n",
       "      <th>ZFAS1</th>\n",
       "      <th>ZFP36L1</th>\n",
       "      <th>ZNF365</th>\n",
       "      <th>ZNF43</th>\n",
       "      <th>ZNF483</th>\n",
       "      <th>ZNF556</th>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGAGTAATC-1</td>\n",
       "      <td>AAACCTGAGAGTAATC</td>\n",
       "      <td>-0.078689</td>\n",
       "      <td>-0.093494</td>\n",
       "      <td>-0.01406</td>\n",
       "      <td>-0.013653</td>\n",
       "      <td>-0.018973</td>\n",
       "      <td>-0.026806</td>\n",
       "      <td>-0.034066</td>\n",
       "      <td>-0.485621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>-1.633805</td>\n",
       "      <td>-0.876912</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGGGATCTG-1</td>\n",
       "      <td>AAACCTGAGGGATCTG</td>\n",
       "      <td>-0.078689</td>\n",
       "      <td>-0.093494</td>\n",
       "      <td>-0.01406</td>\n",
       "      <td>-0.013653</td>\n",
       "      <td>-0.018973</td>\n",
       "      <td>-0.026806</td>\n",
       "      <td>-0.034066</td>\n",
       "      <td>-0.485621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>1.610220</td>\n",
       "      <td>1.322265</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGGTCATCT-1</td>\n",
       "      <td>AAACCTGAGGTCATCT</td>\n",
       "      <td>-0.078689</td>\n",
       "      <td>-0.093494</td>\n",
       "      <td>-0.01406</td>\n",
       "      <td>-0.013653</td>\n",
       "      <td>-0.018973</td>\n",
       "      <td>-0.026806</td>\n",
       "      <td>-0.034066</td>\n",
       "      <td>-0.485621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>-0.567252</td>\n",
       "      <td>0.057540</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGCAATGGAGC-1</td>\n",
       "      <td>AAACCTGCAATGGAGC</td>\n",
       "      <td>-0.078689</td>\n",
       "      <td>-0.093494</td>\n",
       "      <td>-0.01406</td>\n",
       "      <td>-0.013653</td>\n",
       "      <td>-0.018973</td>\n",
       "      <td>-0.026806</td>\n",
       "      <td>-0.034066</td>\n",
       "      <td>2.237599</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>1.765279</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>-1.659420</td>\n",
       "      <td>-0.876912</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCACCAGGCT-1</td>\n",
       "      <td>AAACCTGCACCAGGCT</td>\n",
       "      <td>-0.078689</td>\n",
       "      <td>-0.093494</td>\n",
       "      <td>-0.01406</td>\n",
       "      <td>-0.013653</td>\n",
       "      <td>-0.018973</td>\n",
       "      <td>-0.026806</td>\n",
       "      <td>-0.034066</td>\n",
       "      <td>-0.485621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>0.686753</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>0.541260</td>\n",
       "      <td>0.934539</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>1.406718</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1002 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Metadata_barcode Metadata_sequence     ABCA1     ABCC3   ABI3BP  \\\n",
       "0  AAACCTGAGAGTAATC-1  AAACCTGAGAGTAATC -0.078689 -0.093494 -0.01406   \n",
       "1  AAACCTGAGGGATCTG-1  AAACCTGAGGGATCTG -0.078689 -0.093494 -0.01406   \n",
       "2  AAACCTGAGGTCATCT-1  AAACCTGAGGTCATCT -0.078689 -0.093494 -0.01406   \n",
       "3  AAACCTGCAATGGAGC-1  AAACCTGCAATGGAGC -0.078689 -0.093494 -0.01406   \n",
       "4  AAACCTGCACCAGGCT-1  AAACCTGCACCAGGCT -0.078689 -0.093494 -0.01406   \n",
       "\n",
       "   AC002331.1  AC002480.3  AC003092.1  AC005616.2  AC006262.5  ...    YPEL4  \\\n",
       "0   -0.013653   -0.018973   -0.026806   -0.034066   -0.485621  ... -0.09309   \n",
       "1   -0.013653   -0.018973   -0.026806   -0.034066   -0.485621  ... -0.09309   \n",
       "2   -0.013653   -0.018973   -0.026806   -0.034066   -0.485621  ... -0.09309   \n",
       "3   -0.013653   -0.018973   -0.026806   -0.034066    2.237599  ... -0.09309   \n",
       "4   -0.013653   -0.018973   -0.026806   -0.034066   -0.485621  ... -0.09309   \n",
       "\n",
       "      YPEL5    ZBTB38     ZFAS1   ZFP36L1    ZNF365     ZNF43    ZNF483  \\\n",
       "0 -0.632468 -0.539091 -1.633805 -0.876912 -0.012155 -0.017763 -0.026877   \n",
       "1 -0.632468 -0.539091  1.610220  1.322265 -0.012155 -0.017763 -0.026877   \n",
       "2 -0.632468 -0.539091 -0.567252  0.057540 -0.012155 -0.017763 -0.026877   \n",
       "3  1.765279 -0.539091 -1.659420 -0.876912 -0.012155 -0.017763 -0.026877   \n",
       "4  0.686753 -0.539091  0.541260  0.934539 -0.012155 -0.017763 -0.026877   \n",
       "\n",
       "   ZNF556       ZYX  \n",
       "0 -0.0524 -0.452229  \n",
       "1 -0.0524 -0.452229  \n",
       "2 -0.0524 -0.452229  \n",
       "3 -0.0524 -0.452229  \n",
       "4 -0.0524  1.406718  \n",
       "\n",
       "[5 rows x 1002 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load and process gene expression data\n",
    "gene_exp_file = pathlib.Path(f\"{perturbseq_data_dir}/{gse_id}_processed_matrix.tsv.gz\")\n",
    "\n",
    "gene_exp_df = (\n",
    "    pd.read_csv(gene_exp_file, sep=\"\\t\", index_col=0)\n",
    "    .transpose()\n",
    "    .reset_index()\n",
    "    .rename({\"index\": \"Metadata_barcode\"}, axis=\"columns\")\n",
    ")\n",
    "\n",
    "# Pull out the measured genes\n",
    "gene_features = gene_exp_df.columns.tolist()\n",
    "gene_features.remove(\"Metadata_barcode\")\n",
    "\n",
    "gene_exp_df = gene_exp_df.assign(Metadata_sequence=[x.split(\"-\")[0] for x in gene_exp_df.Metadata_barcode])\n",
    "gene_exp_df.columns.name = \"\"\n",
    "\n",
    "meta_features = [\"Metadata_barcode\", \"Metadata_sequence\"]\n",
    "gene_features = sorted(gene_exp_df.drop(meta_features, axis=\"columns\").columns.tolist())\n",
    "\n",
    "gene_exp_df = gene_exp_df.reindex(meta_features + gene_features, axis=\"columns\")\n",
    "\n",
    "print(gene_exp_df.shape)\n",
    "gene_exp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(23608, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metadata_cell_barcode</th>\n",
       "      <th>Metadata_guide_identity</th>\n",
       "      <th>Metadata_read_count</th>\n",
       "      <th>Metadata_UMI_count</th>\n",
       "      <th>Metadata_coverage</th>\n",
       "      <th>Metadata_gemgroup</th>\n",
       "      <th>Metadata_good_coverage</th>\n",
       "      <th>Metadata_number_of_cells</th>\n",
       "      <th>Metadata_gene_identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GGACAAGTCCCTGACT-3</td>\n",
       "      <td>neg_ctrl_non-targeting_00028</td>\n",
       "      <td>7452</td>\n",
       "      <td>457</td>\n",
       "      <td>16.306346</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>neg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGACTTCAGAAGGCCT-3</td>\n",
       "      <td>GNB2L1_GNB2L1_+_180670873.23-P1P2_13</td>\n",
       "      <td>6554</td>\n",
       "      <td>361</td>\n",
       "      <td>18.155125</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>GNB2L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTAGGCAAGAAGGCCT-2</td>\n",
       "      <td>TUBB_TUBB_+_30688126.23-P1_00</td>\n",
       "      <td>4177</td>\n",
       "      <td>165</td>\n",
       "      <td>25.315152</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>TUBB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGTAGGCAGCCAGGAT-1</td>\n",
       "      <td>TUBB_TUBB_+_30688126.23-P1_01</td>\n",
       "      <td>4024</td>\n",
       "      <td>218</td>\n",
       "      <td>18.458716</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>TUBB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCGCAACTCACGATGT-2</td>\n",
       "      <td>HSPE1_HSPE1_+_198365089.23-P1P2_00</td>\n",
       "      <td>3923</td>\n",
       "      <td>134</td>\n",
       "      <td>29.276119</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>HSPE1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Metadata_cell_barcode               Metadata_guide_identity  \\\n",
       "0    GGACAAGTCCCTGACT-3          neg_ctrl_non-targeting_00028   \n",
       "1    CGACTTCAGAAGGCCT-3  GNB2L1_GNB2L1_+_180670873.23-P1P2_13   \n",
       "2    TTAGGCAAGAAGGCCT-2         TUBB_TUBB_+_30688126.23-P1_00   \n",
       "3    CGTAGGCAGCCAGGAT-1         TUBB_TUBB_+_30688126.23-P1_01   \n",
       "4    GCGCAACTCACGATGT-2    HSPE1_HSPE1_+_198365089.23-P1P2_00   \n",
       "\n",
       "   Metadata_read_count  Metadata_UMI_count  Metadata_coverage  \\\n",
       "0                 7452                 457          16.306346   \n",
       "1                 6554                 361          18.155125   \n",
       "2                 4177                 165          25.315152   \n",
       "3                 4024                 218          18.458716   \n",
       "4                 3923                 134          29.276119   \n",
       "\n",
       "   Metadata_gemgroup  Metadata_good_coverage  Metadata_number_of_cells  \\\n",
       "0                  3                    True                         1   \n",
       "1                  3                    True                         1   \n",
       "2                  2                    True                         2   \n",
       "3                  1                    True                         1   \n",
       "4                  2                    True                         1   \n",
       "\n",
       "  Metadata_gene_identity  \n",
       "0                    neg  \n",
       "1                 GNB2L1  \n",
       "2                   TUBB  \n",
       "3                   TUBB  \n",
       "4                  HSPE1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load cell identities\n",
    "identity_file = pathlib.Path(f\"{perturbseq_data_dir}/{gse_id}_cell_identities.csv.gz\")\n",
    "cell_id_df = pd.read_csv(identity_file, sep=\",\")\n",
    "\n",
    "cell_id_df.columns = [f\"Metadata_{x}\" for x in cell_id_df.columns]\n",
    "cell_id_df = cell_id_df.assign(\n",
    "    Metadata_gene_identity=[str(x).split(\"_\")[0] for x in cell_id_df.Metadata_guide_identity]\n",
    ")\n",
    "\n",
    "print(cell_id_df.shape)\n",
    "cell_id_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(23537, 1012)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metadata_cell_identity</th>\n",
       "      <th>Metadata_cell_barcode</th>\n",
       "      <th>Metadata_guide_identity</th>\n",
       "      <th>Metadata_read_count</th>\n",
       "      <th>Metadata_UMI_count</th>\n",
       "      <th>Metadata_coverage</th>\n",
       "      <th>Metadata_gemgroup</th>\n",
       "      <th>Metadata_good_coverage</th>\n",
       "      <th>Metadata_number_of_cells</th>\n",
       "      <th>Metadata_gene_identity</th>\n",
       "      <th>...</th>\n",
       "      <th>YPEL4</th>\n",
       "      <th>YPEL5</th>\n",
       "      <th>ZBTB38</th>\n",
       "      <th>ZFAS1</th>\n",
       "      <th>ZFP36L1</th>\n",
       "      <th>ZNF365</th>\n",
       "      <th>ZNF43</th>\n",
       "      <th>ZNF483</th>\n",
       "      <th>ZNF556</th>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sc_profile_0</td>\n",
       "      <td>AAACCTGAGAGTAATC-1</td>\n",
       "      <td>RAN_RAN_+_131356438.23-P1P2_12</td>\n",
       "      <td>544.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>16.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RAN</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>-1.633805</td>\n",
       "      <td>-0.876912</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sc_profile_1</td>\n",
       "      <td>AAACCTGAGGGATCTG-1</td>\n",
       "      <td>neg_ctrl_non-targeting_00089</td>\n",
       "      <td>267.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>14.052632</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>neg</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>1.610220</td>\n",
       "      <td>1.322265</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sc_profile_2</td>\n",
       "      <td>AAACCTGAGGTCATCT-1</td>\n",
       "      <td>POLR2H_POLR2H_+_184081251.23-P1P2_08</td>\n",
       "      <td>622.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>18.294118</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>POLR2H</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>-0.567252</td>\n",
       "      <td>0.057540</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sc_profile_3</td>\n",
       "      <td>AAACCTGCAATGGAGC-1</td>\n",
       "      <td>TUBB_TUBB_+_30688126.23-P1_03</td>\n",
       "      <td>433.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>21.650000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TUBB</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>1.765279</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>-1.659420</td>\n",
       "      <td>-0.876912</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sc_profile_4</td>\n",
       "      <td>AAACCTGCACCAGGCT-1</td>\n",
       "      <td>CDC23_CDC23_-_137548987.23-P1P2_04</td>\n",
       "      <td>136.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CDC23</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>0.686753</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>0.541260</td>\n",
       "      <td>0.934539</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>1.406718</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1012 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Metadata_cell_identity Metadata_cell_barcode  \\\n",
       "0           sc_profile_0    AAACCTGAGAGTAATC-1   \n",
       "1           sc_profile_1    AAACCTGAGGGATCTG-1   \n",
       "2           sc_profile_2    AAACCTGAGGTCATCT-1   \n",
       "3           sc_profile_3    AAACCTGCAATGGAGC-1   \n",
       "4           sc_profile_4    AAACCTGCACCAGGCT-1   \n",
       "\n",
       "                Metadata_guide_identity  Metadata_read_count  \\\n",
       "0        RAN_RAN_+_131356438.23-P1P2_12                544.0   \n",
       "1          neg_ctrl_non-targeting_00089                267.0   \n",
       "2  POLR2H_POLR2H_+_184081251.23-P1P2_08                622.0   \n",
       "3         TUBB_TUBB_+_30688126.23-P1_03                433.0   \n",
       "4    CDC23_CDC23_-_137548987.23-P1P2_04                136.0   \n",
       "\n",
       "   Metadata_UMI_count  Metadata_coverage  Metadata_gemgroup  \\\n",
       "0                34.0          16.000000                1.0   \n",
       "1                19.0          14.052632                1.0   \n",
       "2                34.0          18.294118                1.0   \n",
       "3                20.0          21.650000                1.0   \n",
       "4                 8.0          17.000000                1.0   \n",
       "\n",
       "  Metadata_good_coverage  Metadata_number_of_cells Metadata_gene_identity  \\\n",
       "0                   True                       1.0                    RAN   \n",
       "1                   True                       1.0                    neg   \n",
       "2                   True                       1.0                 POLR2H   \n",
       "3                   True                       1.0                   TUBB   \n",
       "4                   True                       1.0                  CDC23   \n",
       "\n",
       "   ...    YPEL4     YPEL5    ZBTB38     ZFAS1   ZFP36L1    ZNF365     ZNF43  \\\n",
       "0  ... -0.09309 -0.632468 -0.539091 -1.633805 -0.876912 -0.012155 -0.017763   \n",
       "1  ... -0.09309 -0.632468 -0.539091  1.610220  1.322265 -0.012155 -0.017763   \n",
       "2  ... -0.09309 -0.632468 -0.539091 -0.567252  0.057540 -0.012155 -0.017763   \n",
       "3  ... -0.09309  1.765279 -0.539091 -1.659420 -0.876912 -0.012155 -0.017763   \n",
       "4  ... -0.09309  0.686753 -0.539091  0.541260  0.934539 -0.012155 -0.017763   \n",
       "\n",
       "     ZNF483  ZNF556       ZYX  \n",
       "0 -0.026877 -0.0524 -0.452229  \n",
       "1 -0.026877 -0.0524 -0.452229  \n",
       "2 -0.026877 -0.0524 -0.452229  \n",
       "3 -0.026877 -0.0524 -0.452229  \n",
       "4 -0.026877 -0.0524  1.406718  \n",
       "\n",
       "[5 rows x 1012 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge single cells with identifiers\n",
    "sc_df = cell_id_df.merge(\n",
    "    gene_exp_df,\n",
    "    how=\"right\",\n",
    "    right_on=\"Metadata_barcode\",\n",
    "    left_on=\"Metadata_cell_barcode\",\n",
    ")\n",
    "\n",
    "sc_df = sc_df.reset_index().rename({\"index\": \"Metadata_cell_identity\"}, axis=\"columns\")\n",
    "sc_df.Metadata_cell_identity = [f\"sc_profile_{x}\" for x in sc_df.Metadata_cell_identity]\n",
    "\n",
    "print(sc_df.shape)\n",
    "sc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the file to disk\n",
    "sc_df.to_csv(output_file, index=False, sep=\"\\t\", compression={\"method\": \"gzip\", \"mtime\": 1})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate bulk perturbseq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(138, 1002)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metadata_guide_identity</th>\n",
       "      <th>Metadata_gene_identity</th>\n",
       "      <th>ABCA1</th>\n",
       "      <th>ABCC3</th>\n",
       "      <th>ABI3BP</th>\n",
       "      <th>AC002331.1</th>\n",
       "      <th>AC002480.3</th>\n",
       "      <th>AC003092.1</th>\n",
       "      <th>AC005616.2</th>\n",
       "      <th>AC006262.5</th>\n",
       "      <th>...</th>\n",
       "      <th>YPEL4</th>\n",
       "      <th>YPEL5</th>\n",
       "      <th>ZBTB38</th>\n",
       "      <th>ZFAS1</th>\n",
       "      <th>ZFP36L1</th>\n",
       "      <th>ZNF365</th>\n",
       "      <th>ZNF43</th>\n",
       "      <th>ZNF483</th>\n",
       "      <th>ZNF556</th>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALDOA_ALDOA_+_30077139.23-P1P2_00</td>\n",
       "      <td>ALDOA</td>\n",
       "      <td>-0.078689</td>\n",
       "      <td>-0.093494</td>\n",
       "      <td>-0.01406</td>\n",
       "      <td>-0.013653</td>\n",
       "      <td>-0.018973</td>\n",
       "      <td>-0.026806</td>\n",
       "      <td>-0.034066</td>\n",
       "      <td>-0.485621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>-0.170790</td>\n",
       "      <td>0.014776</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALDOA_ALDOA_+_30077139.23-P1P2_06</td>\n",
       "      <td>ALDOA</td>\n",
       "      <td>-0.078689</td>\n",
       "      <td>-0.093494</td>\n",
       "      <td>-0.01406</td>\n",
       "      <td>-0.013653</td>\n",
       "      <td>-0.018973</td>\n",
       "      <td>-0.026806</td>\n",
       "      <td>-0.034066</td>\n",
       "      <td>-0.485621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>-0.045114</td>\n",
       "      <td>0.042050</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALDOA_ALDOA_+_30077139.23-P1P2_07</td>\n",
       "      <td>ALDOA</td>\n",
       "      <td>-0.078689</td>\n",
       "      <td>-0.093494</td>\n",
       "      <td>-0.01406</td>\n",
       "      <td>-0.013653</td>\n",
       "      <td>-0.018973</td>\n",
       "      <td>-0.026806</td>\n",
       "      <td>-0.034066</td>\n",
       "      <td>-0.485621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>-0.101183</td>\n",
       "      <td>0.027887</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALDOA_ALDOA_+_30077139.23-P1P2_13</td>\n",
       "      <td>ALDOA</td>\n",
       "      <td>-0.078689</td>\n",
       "      <td>-0.093494</td>\n",
       "      <td>-0.01406</td>\n",
       "      <td>-0.013653</td>\n",
       "      <td>-0.018973</td>\n",
       "      <td>-0.026806</td>\n",
       "      <td>-0.034066</td>\n",
       "      <td>-0.485621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>-0.050848</td>\n",
       "      <td>-0.107389</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ALDOA_ALDOA_+_30077139.23-P1P2_14</td>\n",
       "      <td>ALDOA</td>\n",
       "      <td>-0.078689</td>\n",
       "      <td>-0.093494</td>\n",
       "      <td>-0.01406</td>\n",
       "      <td>-0.013653</td>\n",
       "      <td>-0.018973</td>\n",
       "      <td>-0.026806</td>\n",
       "      <td>-0.034066</td>\n",
       "      <td>-0.485621</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.09309</td>\n",
       "      <td>-0.632468</td>\n",
       "      <td>-0.539091</td>\n",
       "      <td>0.150701</td>\n",
       "      <td>0.051095</td>\n",
       "      <td>-0.012155</td>\n",
       "      <td>-0.017763</td>\n",
       "      <td>-0.026877</td>\n",
       "      <td>-0.0524</td>\n",
       "      <td>-0.452229</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 1002 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Metadata_guide_identity Metadata_gene_identity     ABCA1  \\\n",
       "1  ALDOA_ALDOA_+_30077139.23-P1P2_00                  ALDOA -0.078689   \n",
       "2  ALDOA_ALDOA_+_30077139.23-P1P2_06                  ALDOA -0.078689   \n",
       "3  ALDOA_ALDOA_+_30077139.23-P1P2_07                  ALDOA -0.078689   \n",
       "4  ALDOA_ALDOA_+_30077139.23-P1P2_13                  ALDOA -0.078689   \n",
       "5  ALDOA_ALDOA_+_30077139.23-P1P2_14                  ALDOA -0.078689   \n",
       "\n",
       "      ABCC3   ABI3BP  AC002331.1  AC002480.3  AC003092.1  AC005616.2  \\\n",
       "1 -0.093494 -0.01406   -0.013653   -0.018973   -0.026806   -0.034066   \n",
       "2 -0.093494 -0.01406   -0.013653   -0.018973   -0.026806   -0.034066   \n",
       "3 -0.093494 -0.01406   -0.013653   -0.018973   -0.026806   -0.034066   \n",
       "4 -0.093494 -0.01406   -0.013653   -0.018973   -0.026806   -0.034066   \n",
       "5 -0.093494 -0.01406   -0.013653   -0.018973   -0.026806   -0.034066   \n",
       "\n",
       "   AC006262.5  ...    YPEL4     YPEL5    ZBTB38     ZFAS1   ZFP36L1    ZNF365  \\\n",
       "1   -0.485621  ... -0.09309 -0.632468 -0.539091 -0.170790  0.014776 -0.012155   \n",
       "2   -0.485621  ... -0.09309 -0.632468 -0.539091 -0.045114  0.042050 -0.012155   \n",
       "3   -0.485621  ... -0.09309 -0.632468 -0.539091 -0.101183  0.027887 -0.012155   \n",
       "4   -0.485621  ... -0.09309 -0.632468 -0.539091 -0.050848 -0.107389 -0.012155   \n",
       "5   -0.485621  ... -0.09309 -0.632468 -0.539091  0.150701  0.051095 -0.012155   \n",
       "\n",
       "      ZNF43    ZNF483  ZNF556       ZYX  \n",
       "1 -0.017763 -0.026877 -0.0524 -0.452229  \n",
       "2 -0.017763 -0.026877 -0.0524 -0.452229  \n",
       "3 -0.017763 -0.026877 -0.0524 -0.452229  \n",
       "4 -0.017763 -0.026877 -0.0524 -0.452229  \n",
       "5 -0.017763 -0.026877 -0.0524 -0.452229  \n",
       "\n",
       "[5 rows x 1002 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Perform single cell aggregation into bulk\n",
    "bulk_df = aggregate(\n",
    "    population_df=sc_df,\n",
    "    strata=[\"Metadata_guide_identity\"],\n",
    "    features=gene_features,\n",
    "    operation=\"median\",\n",
    ")\n",
    "\n",
    "# remove one row with NaN value\n",
    "bulk_df = bulk_df[~bulk_df[\"Metadata_guide_identity\"].isnull()]\n",
    "\n",
    "# create a column for the gene\n",
    "bulk_df = bulk_df.assign(Metadata_gene_identity=[x.split(\"_\")[0] for x in bulk_df.Metadata_guide_identity]).query(\n",
    "    \"Metadata_gene_identity != '*'\"\n",
    ")\n",
    "\n",
    "bulk_df = bulk_df.reindex(\n",
    "    [\"Metadata_guide_identity\", \"Metadata_gene_identity\"] + gene_features,\n",
    "    axis=\"columns\",\n",
    ")\n",
    "\n",
    "print(bulk_df.shape)\n",
    "bulk_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the file to disk\n",
    "bulk_df.to_csv(output_bulk_file, index=False, sep=\"\\t\", compression={\"method\": \"gzip\", \"mtime\": 1})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "grit-benchmark",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "vscode": {
   "interpreter": {
    "hash": "5263a8d172dade9e46e5af2db82be6add0b3c649f4dab478dd2c42518a050092"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
